### DISCUSSION TITLE
correct define to use on Windows only code

### DISCUSSION THREAD

**Discussion by dimitre** [openframeworks/openFrameworks#7450] [General]
I've seen in different places of OF a define called TARGET_WIN32
like
#if defined(TARGET_WIN32)

is this true for all windows platform, including 64 bits?
should we use it or instead OF_TARGET_WINDOWS or something similar? 
or maybe even better, have 

OF_PLATFORM_WINDOWS
OF_PLATFORM_MACOS
OF_PLATFORM_LINUX

to define pieces of code that will run in all targets of the same platform?
thoughts?

**Comment by artificiel** [comment]
to answer the first "direct" question: `TARGET_WIN32` is defined in ofConstant.h if [`#ifdef _WIN32`](https://github.com/openframeworks/openFrameworks/blob/1e77cec5465446b91ff21cc915aa8d646de98099/libs/openFrameworks/utils/ofConstants.h#L88) , and here it stipulates that [`_WIN32` includes 64 bits](https://sourceforge.net/p/predef/wiki/OperatingSystems/).

that being said there is a bit of chaos in [ofConstants.h](https://github.com/openframeworks/openFrameworks/blob/master/libs/openFrameworks/utils/ofConstants.h) it seems the logic of the definitions is not consistent... there is an effort like your `OF_PLATFORM_*` idea around `TARGET_OF_*`, but windows is immediately skipping the rule with `TARGET_WIN32`.  also there is `TARGET_OF_IPHONE` and `TARGET_OF_IOS` (Core seems to use only `TARGET_OF_IOS` but there are some addons that use `TARGET_OF_IPHONE`).

so to answer the second question it seems the logical thing is to support the effort done in ofConstant to normalize `TARGET_OF_*`, but maybe fix windows to `TARGET_OF_WIN32` (or `TARGET_OF_WINDOWS` as you suggest it is more inclusive and probably "modern") and make sure all the #ifdefs use the same pattern (while keeping the old `TARGET_WIN32` for perennial addon compatibility).
  https://github.com/openframeworks/openFrameworks/blob/1e77cec5465446b91ff21cc915aa8d646de98099/libs/openFrameworks/utils/ofConstants.h#L88-L131

**Reply by dimitre** [reply]
I've converted this to discussion. yes lots of things have to be restructured here.
Maybe separating OS and platform variant in two different kinds of things. A little bit like github actions do.
there you can use like this
```
if: runner.os == 'Linux'

if: runner.os == 'Windows'

if: runner.os == 'macOS'
```

and then you have the TARGET defines for platform variants.
even then there are mixed stuff like target referring different kinds of functionalities
TARGET_OPENGLES
TARGET_RASPBERRY_PI_LEGACY
TARGET_RASPBERRY_PI

**Comment by artificiel** [comment]
the "original" targets as defined (for instance) [by apple](https://opensource.apple.com/source/CarbonHeaders/CarbonHeaders-8A428/TargetConditionals.h.auto.html) are more precise than OF: `TARGET_OS_IOS` means "the OS of the target is iOS). `TARGET_OF_IOS` loses a distinguishing layer. as I understand the pattern the idea is to lump defines in functionally useful units. so you would rename `TARGET_OPENGLES` to something like `TARGET_RENDERER_OPENGLES`? or even `OF_TARGET_RENDERER_OPENGLES`? `OF_TARGET_OS_IOS`? 

concretely, nomenclature drift is at play but (apart from clarifying WIN32 includes 64) are there specific use-cases that need to be addressed?

**Reply by dimitre** [reply]
haha yeah this time it is me trying to think some steps ahead. 
For now we can have a definition of a meaningful windows constant to be used in the core.
the hard thing is how to decide something to be in parallel with all the historical mess.
for ex. I personaly like the enum that begins with OF_ , like for ex: OF_TARGET_OSX
what I like about it, it explicitely begins with OF so with a quick glance you know it is not a compiler or platform define.
and it doesn't get too similar like in TARGET_OS_IOS and TARGET_OF_IOS.

but it can get confusing mixing defines with enum.
if it is something that can be ignored, my suggestion would be

OF_OS_MACOS
OF_OS_WINDOWS
OF_OS_LINUX
OF_OS_IOS

ah yes, and it doesn't use target when mentioning OS too

**Comment by ofTheo** [comment]
I'd totally be up for standardizing these ( as long we kept the legacy defines for a while after ). 
Having sub defines for VS vs MINGW/MSYS and IPHONE / IPAD etc makes sense. 

```
OF_OS_EMSCRIPTEN
OF_OS_MACOS
OF_OS_WINDOWS
  OF_TOOLCHAIN_VS
  OF_TOOLCHAIN_MSYS
OF_OS_LINUX
OF_OS_IOS
  OF_DEVICE_IPHONE
  OF_DEVICE_IPAD 
```

Also like:
```
OF_RENDERER_OPENGLES
OF_RENDERER_OPENGL
OF_RENDERER_CAIRO 
```
etc

**Reply by dimitre** [reply]
Yess love this definitions. I can work on that soon

**Comment by artificiel** [comment]
about `OF_DEVICE_IPHONE` I don't think there is a way to figure that out at compile time (short of scanning  the Xcode project?). Apple does not define that (below the relevant section of TargetConditionals as of 13.1).

also I'm wondering about the enum `ofTargetPlatform` and the companion function `ofGetTargetPlatform()` -- it seems to be only used very sparsely around MINGW -- I don't know about MINGW and maybe something special happens at runtime, but it may be something that could be simply `#ifdef'd`? (so we could remove this sort of "shadow" of the defines).

that makes me wonder about the actual usage -- perhaps an audit of the `#ifdef` would be interesting to see the actual usage, and make sure we don't duplicate things for nothing, as in all cases libraries will be looking for the conventional `#defines`, the `OF_*` stuff has to facilitate specific OF needs.


```
 *  TARGET_OS_*
 *  These conditionals specify in which Operating System the generated code will
 *  run.  Indention is used to show which conditionals are evolutionary subclasses.
 *
 *  The MAC/WIN32/UNIX conditionals are mutually exclusive.
 *  The IOS/TV/WATCH conditionals are mutually exclusive.
 *
 *
 *      TARGET_OS_WIN32           - Generated code will run under WIN32 API
 *      TARGET_OS_WINDOWS         - Generated code will run under Windows
 *      TARGET_OS_UNIX            - Generated code will run under some Unix (not OSX)
 *      TARGET_OS_LINUX           - Generated code will run under Linux
 *      TARGET_OS_MAC             - Generated code will run under Mac OS X variant
 *         TARGET_OS_OSX          - Generated code will run under OS X devices
 *         TARGET_OS_IPHONE          - Generated code for firmware, devices, or simulator
 *            TARGET_OS_IOS             - Generated code will run under iOS
 *            TARGET_OS_TV              - Generated code will run under Apple TV OS
 *            TARGET_OS_WATCH           - Generated code will run under Apple Watch OS
 *            TARGET_OS_BRIDGE          - Generated code will run under Bridge devices
 *            TARGET_OS_MACCATALYST     - Generated code will run under macOS
 *         TARGET_OS_DRIVERKIT          - Generated code will run under macOS, iOS, Apple TV OS, or Apple Watch OS
 *         TARGET_OS_SIMULATOR      - Generated code will run under a simulator
 *
 *      TARGET_OS_EMBEDDED        - DEPRECATED: Use TARGET_OS_IPHONE and/or TARGET_OS_SIMULATOR instead
 *      TARGET_IPHONE_SIMULATOR   - DEPRECATED: Same as TARGET_OS_SIMULATOR
 *      TARGET_OS_NANO            - DEPRECATED: Same as TARGET_OS_WATCH
 *
 *    +---------------------------------------------------------------------------+
 *    |                             TARGET_OS_MAC                                 |
 *    | +-----+ +-------------------------------------------------+ +-----------+ |
 *    | |     | |                  TARGET_OS_IPHONE               | |           | |
 *    | |     | | +-----------------+ +----+ +-------+ +--------+ | |           | |
 *    | |     | | |       IOS       | |    | |       | |        | | |           | |
 *    | | OSX | | | +-------------+ | | TV | | WATCH | | BRIDGE | | | DRIVERKIT | |
 *    | |     | | | | MACCATALYST | | |    | |       | |        | | |           | |
 *    | |     | | | +-------------+ | |    | |       | |        | | |           | |
 *    | |     | | +-----------------+ +----+ +-------+ +--------+ | |           | |
 *    | +-----+ +-------------------------------------------------+ +-----------+ |
 *    +---------------------------------------------------------------------------+

**Reply by dimitre** [reply]
Yes, ofGetTargetPlatform() is not used much in the core, but it is being used more in projectGenerator. 
There is a lot of potential for simplification in projectGenerator, it is already a lot less complex than it was one or two years ago.
I'll be re-checking this functionality there soon
